projects/gloo/pkg/plugins/adaptive_concurrency/plugin.go

```diff
@@ -0,0 +1,221 @@
+	ErrMinRttCalcParamsMissing                = fmt.Errorf("min_rtt_calc_params is required")
+	ErrIntervalMissing                        = fmt.Errorf("Either interval_millis or fixed_value_millis must be set")
+)
+
+const (
+	ExtensionName = "envoy.extensions.filters.http.adaptive_concurrency.v3.AdaptiveConcurrency"
+
+	// Default values as documented in the proto file
+	// These match the defaults in the envoy docs: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto#envoy-v3-api-msg-extensions-filters-http-adaptive-concurrency-v3-adaptiveconcurrency
+	// These are set in the plugin to avoid drift in our documentation in the Envoy values change.
+	DefaultSampleAggregatePercentile      = 50.0
+	DefaultMaxConcurrencyLimit            = 1000
+	DefaultRequestCount                   = 50
+	DefaultMinConcurrency                 = 3
+	DefaultJitterPercentile               = 15.0
+	DefaultBufferPercentile               = 25.0
+	DefaultConcurrencyLimitExceededStatus = 503
+)
+
+var (
+	_ plugins.Plugin           = new(plugin)
+	_ plugins.HttpFilterPlugin = new(plugin)
+)
+
+type plugin struct{}
+
+func NewPlugin() *plugin {
+	return &plugin{}
+}
+func (p *plugin) Name() string {
+	return ExtensionName
+}
+
+func (p *plugin) Init(params plugins.InitParams) {}
+
+func (p *plugin) HttpFilters(params plugins.Params, listener *v1.HttpListener) ([]plugins.StagedHttpFilter, error) {
+	in := listener.GetOptions()
+	adaptiveConcurrencyConfig, err := translateAdaptiveConcurrency(in)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if adaptiveConcurrencyConfig == nil {
+		return []plugins.StagedHttpFilter{}, nil
+	}
+
+	return []plugins.StagedHttpFilter{plugins.MustNewStagedFilter(ExtensionName, adaptiveConcurrencyConfig, pluginStage)}, nil
+}
+
+func translateAdaptiveConcurrency(in *v1.HttpListenerOptions) (*envoy_adaptive_concurrency_v3.AdaptiveConcurrency, error) {
+	adaptiveConcurrency := in.GetAdaptiveConcurrency()
+
+	if adaptiveConcurrency == nil {
+		return nil, nil
+	}
+
+	concurrencyLimitParams, err := translateConcurrencyLimitParams(adaptiveConcurrency)
+	if err != nil {
+		return nil, err
+	}
+
+	minRttCalcParams, err := translateMinRttCalcParams(adaptiveConcurrency.GetMinRttCalcParams())
+	if err != nil {
+		return nil, err
+	}
+
+	// Variable declarations with defaults
+	var (
+		sampleAggregatePercentile      float64
+		concurrencyLimitExceededStatus uint32
+	)
+
+	// Set sample_aggregate_percentile with default
+	if adaptiveConcurrency.GetSampleAggregatePercentile() != nil {
+		sampleAggregatePercentile = adaptiveConcurrency.GetSampleAggregatePercentile().GetValue()
+	} else {
+		sampleAggregatePercentile = DefaultSampleAggregatePercentile
+	}
+
+	// Set concurrency limit exceeded status with default of 503 for non-error codes
+	status := adaptiveConcurrency.GetConcurrencyLimitExceededStatus()
+	if status == 0 || status < 400 {
```

Comment:
Should this be < 500 ? Since this is not a client error ?
Also log an error if so ?